### ISSUE TITLE
Animations are not drawn

### ISSUE THREAD

**Issue by daell82** [eclipse-gef/gef-classic#376]
Hello, 

after looking into the flow example again, I've noticed that the animations for the graph layout are not drawn anymore. I was able to track it down to the changes introduced in #127. This also affects the animations for palette drawers and animated zoom. 

Currently, for me this is reproducible on Windows. On Linux I am not sure whether this was ever supported.

**Comment by ptziegler** [comment]
Can you perhaps post a screenshot of what you mean?

I've reverted the change locally, but I couldn't see any differences. Though that might partially be due to me using Linux, which you've already hinted at.

**Comment by daell82** [comment]
Yes, for the recording I've duplicated the drawer in the palette, otherwise it is not collapsible.

When the `canvas.update()` is omitted, the animations are still processed for their delay but not drawn, which also feels a little bit "laggy". 


https://github.com/eclipse/gef-classic/assets/139544371/3eb7f1d5-13d1-4df6-b6c9-32a4f9505603

**Comment by ptziegler** [comment]
Ah, so that what you mean. Thanks for the example!

Hm, it doesn't look like those animations are played at all on Linux, regardless of whether the update method is called...
I have to dig up my old Windows laptop to see what's going on here.

**Comment by daell82** [comment]
No problem.

On the other hand, I was not able to reproduce the flickering issues mentioned in #127 on Windows. Maybe it could be a quick solution, if the `canvas.update()` is only performed on that platform.

**Comment by azoitl** [comment]
I also noticed that with the improvements of SWT we sometimes see improvements in Draw2d too. 

@Phillipus can you please comment here as well?

**Comment by Phillipus** [comment]
I don't think the changes introduced in https://github.com/eclipse/gef-classic/pull/127 have anything to do with animations. What I found was that SWT needs time to call `Display.getDefault().readAndDispatch()` to update the animation. See https://github.com/archimatetool/archi/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/Animation.java

**Comment by Phillipus** [comment]
Actually, it seems that reverting #127 allows animation to work on Windows, but animation is not working on Mac and Linux which is why I modified the `Animation` class. So, even if #127 it won't fix the problem on all platforms.

**Comment by Phillipus** [comment]
You could try https://github.com/eclipse/gef-classic/pull/377 to see if that works.

**Comment by daell82** [comment]
That change fixes the palette animations and the animated zoom (which I use from GMF in my application). 👍 

For the flow-example the graph layout uses it's own static animation constructs. When I add `Display.getDefault().readAndDispatch();` to the `static boolean playbackState(IFigure container)` method in GraphAnimation.java@146 before the last return statement, it is animated as well.

**Comment by daell82** [comment]
Yes, it does, but having animations on Windows and flickering without animations on other platforms is somewhat undesirable..

I've opened a PR #378 which puts the `Display.readAndDispatch` into the `DeferredUpdateManager.performUpdate`. This shows animations on Windows and on Linux. But I am not that deep into the SWT-drawing mechanisms, therefore I can't estimate, whether this introduces other side effects, The flickering issues didn't appear.

When looking through the repository, the `UpdateManager.performUpdate()` is mainly used during animations and a few other things like smooth scrolling.

**Comment by Phillipus** [comment]
> which puts the Display.readAndDispatch into the DeferredUpdateManager.performUpdate

I thought about that but as the problem is related to animation I put it in the `Animation` class.

**Comment by daell82** [comment]
I guess, that's a little bit complex 😄 

It got my attention when when I noticed that the animations on Windows were not drawn after #127 and when checking it on Linux, they didn't appear even with the revert. Therefore my guess was, that this wasn't supported even before. With the fix only in the `Animation` class, it does not cover all features (e.g. the animated zoom from GMF and the graph layout from the flow example, probably even more) as they do their own logic. Since #127 is a fix for non-Windows platforms dealing with a Windows-issue, I wouldn't revert it back.

**Comment by Phillipus** [comment]
> Since https://github.com/eclipse/gef-classic/pull/127 is a fix for non-Windows platforms dealing with a Windows-issue, I wouldn't revert it back.

Actually, it only affects Windows so an OS check could be applied there. However, it would still break animation on Windows.

**Comment by daell82** [comment]
Ok, that is strange.

I just checked the master branch on a Linux system. There are no animations drawn, neither in the palette, nor in the flow-example graph layout.

**Comment by Phillipus** [comment]
> Ok, that is strange.
> 
> I just checked the master branch on a Linux system. There are no animations drawn, neither in the palette, nor in the flow-example graph layout.
 
Do you mean with the `Display.readAndDispatch` patch? It's working for me on Linux X11. Maybe you use Wayland?

**Comment by daell82** [comment]
With the current master branch, without any modifications, there are no animations. With `Display.readAndDispatch` in the `Animation` class, the palette gets animated. Instead when putting it into the `DeferredUpdateManager` everything seems ok.

I'm testing with OpenSuse in Hyper-V, it says X11.

**Comment by Phillipus** [comment]
To be clear:

> Animations used to work on Linux and Mac up to about 2017 or 2018.

**Comment by daell82** [comment]
Could that be related to this issue:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=481200 ?

**Comment by Phillipus** [comment]
> Could that be related to this issue: https://bugs.eclipse.org/bugs/show_bug.cgi?id=481200 ?

That's my bug report.

There's no way of determining what caused animations to stop working on Mac and Linux. I think probably that changes to the OS were the cause, not SWT.

**Comment by daell82** [comment]
Oh, I didn't see that 🙈 

If I could, I would test it on OSX. But I can delegate it tomorrow if appreciated.

**Comment by Phillipus** [comment]
To be clear, Animations **do** work on Linux and Mac with the `Display.readAndDispatch` workaround. As I said earlier, it's what we use in our RCP app.

**Comment by ptziegler** [comment]
From what I can tell, the issue is caused by the following contract of the UpdateManager no longer being  satisfied:

```
/**
 * Forces an update to occur. Update managers will perform updates
 * automatically, but may do so asynchronously. Calling this method forces a
 * synchronous update.
 */
public abstract void performUpdate();
```

This issue only shows up when the FigureCanvas is created using the `SWT.DOUBLE_BUFFERED` flag because then, the [NativeGraphicsSource](https://github.com/eclipse/gef-classic/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/NativeGraphicsSource.java) is used internally.

However, because the call to `update()` has been removed with #127, this update is now done asychronously. For example, the `redraw()` method (which is still executed) is using [RedrawWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-redrawwindow) on Windows.

And according to the documentation:

> The following flags control when repainting occurs. RedrawWindow will not repaint unless one of these flags is specified:
> - RDW_ERASENOW
> - RDW_UPDATENOW

Though reverting #127 is not the solution as it only works on Windows, and because it paints too much (as shown by the artifacts of the originating bug).

I tried out a few things and this is what I learned:

1) The "original" behavior by using `canvas.update()`.

https://github.com/eclipse/gef-classic/assets/70652978/55604019-1142-4f7c-89c1-20d3c87a407c

2) Instead of returning `null`, the `getGraphics()` method returns a proper `Graphics` object.

https://github.com/eclipse/gef-classic/assets/70652978/93328682-add6-4d28-ade4-e6eeb6e741d4

3) Instead of calling `canvas.update()`, I now call `Display.readAndDispatch()` directly after the call to `redraw()`.

https://github.com/eclipse/gef-classic/assets/70652978/0e34599c-c147-4ea4-9043-ca676d971140

Both solutions 2) and 3) work, but 2) causes a lot of flickering. I'm not overly excited with 3), as there is no guarantee that the event that is executed is exactly the redraw event. But given that it worked in Archi for years, I think that' the best bet.

**Comment by daell82** [comment]
Thank you for also deep diving into this.

I can verify that the third solution seems to fix all issues I observed.

**Comment by ptziegler** [comment]
This one has been going on for several months now... @azoitl Should we include one of the fixes before the M1 and finally get this sorted out?

**Comment by azoitl** [comment]
@ptziegler yes we should. I'm still not perfectly sure which one. But I have a slight tendency to your solution as it is the more general solution. Am I right?

**Comment by ptziegler** [comment]
It's the only one that doesn't cause test failures 😅

**Comment by Phillipus** [comment]
But as the problem is only to do with animation changing `NativeGraphicsSource` seems a bit too much perhaps?

**Comment by azoitl** [comment]
Oh I was to fast and already merged. Should I revert, or should we give it a try?  Early in the development cycle I tend to keep it.

**Comment by ptziegler** [comment]
I still believe that the issue stems from repaint operation triggered by `performUpdate()` being executed asynchronously, going directly against the documentation.

This issue is very noticeable with the animations, but it would generally happen whenever paint operations are done consecutively, as only the "last" repaint is processed. 

Regarding the other proposals: I'm still not sure what's causing the `Argument not valid` exceptions in https://github.com/eclipse/gef-classic/pull/377, but my current guess is that, given that the call `readAndDispatch()` might break the order in which certain UI tasks have to be executed.

The fact that we both had to guard against this case in the Animation class is an indicator for this.

**Comment by ptziegler** [comment]
Urgh... the test failures seem to be a result of this. Because the animations are drawn again, adding e.g. an edit part takes longer. The result is that the animation isn't finished by the time SWTBot checks the new state of the editor.

We probably should've done a rebase before merging. Given that M1 is due today, I would suggest rolling the commit back until this is sorted out.

**Comment by ptziegler** [comment]
Reopened because ee7f7eed4ed1544bd4ddc4ad92e8f3986e939947 was reverted.

**Comment by azoitl** [comment]
Then I'll start M1 now and we see how this can be fixed later.

**Comment by daell82** [comment]
Hello again, 
Thank you again for your effort. I guess the assumption that the solution with `readAndDispatch()` breaks some events is correct. I've noticed this in my application on the `PaletteScrollBar` where the scroll-buttons use the `ButtonModel#REPEAT_FIRING_BEHAVIOR` which causes the Palette to become nearly unusable. In that case the Timer for firing events even remains active after the editor is closed. I've created recordings on Windows and Linux to visualize this with the FlowExample (by overloading PaletteDrawer):


https://github.com/eclipse/gef-classic/assets/139544371/9d815b40-d413-45da-b1ee-d8b0867eb52b

On Linux, there is rarely a dancing palette, but performance while animating afterwards decreases

https://github.com/eclipse/gef-classic/assets/139544371/8c86c6b4-26ed-42f5-a255-f931a215fd26

On Windows it's obvious

I assume that this issue is becoming more complex, since not all parts use the Animation class for animating their contents (e.g. the `GraphAnimation` from the flow example). I'm circumventing this currently by copying the PaletteScrollBar and using the `ButtonModel#DEFAULT_FIRING_BEHAVIOR` , since it's the only situation I've noticed this.

**Comment by ptziegler** [comment]
> Thank you again for your effort. I guess the assumption that the solution with readAndDispatch() breaks some events is correct. I've noticed this in my application on the PaletteScrollBar where the scroll-buttons use the ButtonModel#REPEAT_FIRING_BEHAVIOR which causes the Palette to become nearly unusable. 

To avoid any misunderstandings, _which_ of the three proposed fixes are you using? Given that they are applied to different locations, each one may introduce their own set of unwanted side-effects.

**Comment by daell82** [comment]
I am using the third variant with `canvas.getDisplay().readAndDispatch()` in the `NativeGraphicsSource`. I have no other modifications in the repository (except overloading the PaletteDrawer from the FlowExample for the recordings).
